stan/math/rev/fun/squared_distance.hpp

```diff
@@ -62,29 +58,23 @@ class squared_distance_vv_vari : public vari {
   vari** v2_;
   size_t length_;
 
-  template <int R1, int C1, int R2, int C2>
-  inline static double var_squared_distance(
-      const Eigen::Matrix<var, R1, C1>& v1,
-      const Eigen::Matrix<var, R2, C2>& v2) {
-    using idx_t = index_type_t<matrix_v>;
-
-    return (Eigen::Ref<const vector_v>(v1).val()
-            - Eigen::Ref<const vector_v>(v2).val())
-        .squaredNorm();
-  }
-
  public:
-  template <int R1, int C1, int R2, int C2>
-  squared_distance_vv_vari(const Eigen::Matrix<var, R1, C1>& v1,
-                           const Eigen::Matrix<var, R2, C2>& v2)
-      : vari(var_squared_distance(v1, v2)), length_(v1.size()) {
+  template <
+      typename EigVecVar1, typename EigVecVar2,
+      require_all_eigen_vector_vt<is_var, EigVecVar1, EigVecVar2>* = nullptr>
+  squared_distance_vv_vari(const EigVecVar1& v1, const EigVecVar2& v2)
+      : vari((as_column_vector_or_scalar(v1).val()
+              - as_column_vector_or_scalar(v2).val())
+                 .squaredNorm()),
+        length_(v1.size()) {
     v1_ = reinterpret_cast<vari**>(
         ChainableStack::instance_->memalloc_.alloc(length_ * sizeof(vari*)));
     v2_ = reinterpret_cast<vari**>(
         ChainableStack::instance_->memalloc_.alloc(length_ * sizeof(vari*)));
```

Comment:
[optional]

I think these can be replaced by alloc_array<vari*>

Reply:
As I said earlier I am already changing quite a few things at once. If we do too much at once there is larger chance that something is missed. So I would leave these for a separate series of PRs.